File: iosApp/Well/SwiftUI/Screens/Experts/Filter/FilterScreen.swift
//
//  FilterScreen.swift
//  Well
//
//  Created by Phil on 06.05.2021.
//  Copyright © 2021 Well. All rights reserved.
//

import SharedMobile
import SwiftUI

struct FilterScreen: View {
    let state: FilterFeature.State
    let listener: (FilterFeature.Msg) -> Void
    let hide: () -> Void

    var body: some View {
        NavigationBar(title: HStack {
            Image(systemName: "line.horizontal.3.decrease.circle")
            Text("Filter")
        })
        ScrollView {
            HStack {
                Text("Sort by")
                SelectableHVStack(
                    items: UsersFilter.SortByCompanion().allCases,
                    selectedIndex: .init(get: {
                        Int(state.sortByIndex)
                    }, set: {
                        listener(FilterFeature.MsgSetSortByIndex(index: Int32($0)))
                    })
                ) { item, _ in
                    Text("\(item.name)")
                        .padding(.vertical, 5)
                        .padding(.horizontal, 10)
                }
            }.style(.body1Light).padding()
            Divider()
            ForEach(state.fields, id: \.self) { field in
                EditingField1(field, listener: listener)
                    .padding(.vertical, 7).padding(.horizontal)
                Divider()
            }
            HStack {
                Text("Rating")
                    .style(.body1).padding()
                Spacer()
            }
            SelectableHVStack(
                items: UsersFilter.RatingCompanion().allCases,
                selectedIndex: .init(get: {
                    Int(state.ratingIndex)
                }, set: {
                    listener(FilterFeature.MsgSetRatingIndex(index: Int32($0)))
                })
            ) { item, selected in
                HStack {
                    Text("\(item.title)")
                        .style(.body1Light)
                    if item != UsersFilter.Rating.all {
                        Image(systemName: "star.fill")
                            .foregroundColorKMM(selected ? ColorConstants.White : ColorConstants.LightGray)
                    }
                }
                    .padding(.vertical, 5)
                    .padding(.horizontal, 10)
            }.padding()
            HStack {
                Toggle(isOn: .init(get: {
                    state.filter.withReviews
                }, set: { _ in
                    listener(FilterFeature.MsgToggleWithReviews())
                })) {
                    Text("With reviews")
                }
            }.padding()
            Button {
                listener(FilterFeature.MsgShow())
                hide()
            } label: {
                Text("Show")
            }.buttonStyle(ActionButtonStyle(style: .onWhite))
            Button {
                listener(FilterFeature.MsgClear())
            } label: {
                Text("clear all")
            }.buttonStyle(ActionButtonStyle(style: .white))
        }
    }
}

private struct EditingField1<Msg: AnyObject>: View {
    let field: UIEditingField<UIEditingFieldContentList<AnyObject>, Msg>
    let listener: (Msg) -> Void
    @State
    private var text: String
    @State private var showModal = false

    init(_ field: UIEditingField<UIEditingFieldContentList<AnyObject>, Msg>, listener: @escaping (Msg) -> Void) {
        self.field = field
        self.listener = listener

        _text = State(initialValue: field.content.description)
    }

    var body: some View {
        Button {
            showModal = true
        } label: {
            HStack {
                Text(field.placeholder)
                    .foregroundColorKMM(ColorConstants.Black)
                Spacer()
                Image(systemName: "chevron.right")
                    .foregroundColorKMM(ColorConstants.LightGray)
            }.style(.body1)
                .padding(.vertical, 7)
        }
            .sheet(isPresented: $showModal) {
                selectionScreen()
            }
    }

    @ViewBuilder
    private func selectionScreen() -> some View {
        FilterSelectionScreen(
            title: field.placeholder,
            selection: field.content.selectionIndices,
            variants: field.content.itemDescriptions,
            multipleSelection: field.content.multipleSelectionAvailable,
            onSelectionChanged: { newSelection in
                Napier.i(newSelection, field.content.doCopy(selectionIndices: newSelection))
                listener(
                    field.updateMsg(
                        field.content.doCopy(selectionIndices: newSelection)
                    )!
                )
                showModal = false
            },
            onBack: {
                showModal = false
            }
        )
    }
}

private struct FilterSelectionScreen: View {
    let title: String
    @State var selection: Set<KotlinInt>
    let variants: [String]
    let multipleSelection: Bool
    let onSelectionChanged: (Set<KotlinInt>) -> Void
    let onBack: () -> Void

    var body: some View {
        NavigationBar(
            title: title,
            leftItem: NavigationBarItem(view: Image(systemName: "chevron.left")) {
                onSelectionChanged(selection)
            },
            rightItem: { () -> NavigationBarItem<Text>? in
                guard multipleSelection else {
                    return nil
                }
                let allVariants = Set(Array(variants.indices).map(KotlinInt.init(integerLiteral:)))
                let allVariantsSelected = selection == allVariants
                    return NavigationBarItem(text: allVariantsSelected ? "Deselect all" : "Select all") {
                    selection = allVariantsSelected ? Set() : allVariants
                }
            }()
        )
        SelectionScreen(
            selection: $selection,
            variants: variants,
            multipleSelection: multipleSelection,
            onSelectionChanged: onSelectionChanged
        )
    }
}

File: modules/androidUi/src/main/kotlin/com/well/modules/androidUi/composableScreens/experts/filter/FilterScreen.kt
package com.well.modules.androidUi.composableScreens.experts.filter

import com.well.modules.androidUi.theme.body1Light
import com.well.modules.androidUi.customViews.ActionButton
import com.well.modules.androidUi.customViews.ActionButtonStyle
import com.well.modules.androidUi.customViews.Control
import com.well.modules.androidUi.customViews.ControlItem
import com.well.modules.androidUi.customViews.NavigationBar
import com.well.modules.androidUi.customViews.toNavigationTitleText
import com.well.modules.androidUi.ext.backgroundKMM
import com.well.modules.androidUi.ext.borderKMM
import com.well.modules.androidUi.ext.toColor
import com.well.modules.models.Color
import com.well.modules.models.UsersFilter
import com.well.modules.features.experts.expertsFeature.filter.FilterFeature.Msg
import com.well.modules.features.experts.expertsFeature.filter.FilterFeature.State
import com.well.modules.utils.viewUtils.UIEditingField
import android.annotation.SuppressLint
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Divider
import androidx.compose.material.Icon
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Switch
import androidx.compose.material.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowForward
import androidx.compose.material.icons.filled.FilterList
import androidx.compose.material.icons.filled.Star
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.google.accompanist.flowlayout.FlowRow

private typealias EditingField = UIEditingField<UIEditingField.Content.List<*>, Msg.Update>

@Composable
fun ColumnScope.FilterScreen(
    state: State,
    listener: (Msg) -> Unit,
    hide: () -> Unit,
) {
    val selectedFieldState = remember { mutableStateOf<EditingField?>(null) }
    val selectedField = selectedFieldState.value
    if (selectedField == null) {
        FilterScreenContent(
            state = state,
            listener = listener,
            hide = hide,
            selectField = {
                selectedFieldState.value = it
            },
        )
    } else {
        FilterSelectionScreen(
            title = selectedField.placeholder,
            initialSelection = selectedField.content.selectionIndices,
            variants = selectedField.content.itemDescriptions,
            multipleSelection = selectedField.content.multipleSelectionAvailable,
            onFinish = { newSelection ->
                listener(selectedField.updateMsg(selectedField.content.doCopy(selectionIndices = newSelection)))
                selectedFieldState.value = null
            }
        )
    }
}

@Composable
private fun FilterScreenContent(
    state: State,
    listener: (Msg) -> Unit,
    hide: () -> Unit,
    selectField: (EditingField) -> Unit
) {
    NavigationBar(
        title = ControlItem(
            view = {
                Row {
                    Icon(
                        Icons.Default.FilterList,
                        contentDescription = null,
                        tint = Color.White.toColor(),
                    )
                    "Filter".toNavigationTitleText()
                }
            }
        ),
    )
    LazyColumn {
        item {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .padding(horizontal = 20.dp, vertical = 10.dp)
            ) {
                Text(
                    "Sort by",
                    style = MaterialTheme.typography.body1Light,
                )
                Spacer(modifier = Modifier.size(10.dp))
                SelectableRowColumnGrid(
                    items = UsersFilter.SortBy.allCases,
                    selectedIndex = state.sortByIndex,
                    selectIndex = {
                        listener(Msg.SetSortByIndex(it))
                    },
                    itemBuilder = { sortBy, selected ->
                        Text(
                            sortBy.name,
                            style = MaterialTheme.typography.body1Light,
                            color = (if (selected) Color.White else Color.Black).toColor(),
                            modifier = Modifier
                                .selectableRowColumnItemBackground(selected = selected)
                                .padding(vertical = 5.dp, horizontal = 10.dp)
                        )
                    }
                )
            }
        }
        item {
            Divider()
        }
        items(state.fields) { field ->
            Control(onClick = {
                selectField(field)
            }, content = {
                Row(
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier
                        .padding(horizontal = 20.dp, vertical = 10.dp)
                        .fillMaxSize()
                ) {
                    Text(field.placeholder)
                    Icon(
                        Icons.Default.ArrowForward,
                        contentDescription = null,
                        tint = Color.LightGray.toColor()
                    )
                }
            })
            Divider()
        }
        item {
            Column(
                Modifier.padding(horizontal = 20.dp, vertical = 10.dp)
            ) {
                Text(
                    "Rating",
                    style = MaterialTheme.typography.body1
                )
                SelectableRowColumnGrid(
                    items = UsersFilter.Rating.allCases,
                    selectedIndex = state.ratingIndex,
                    selectIndex = {
                        listener(Msg.SetRatingIndex(it))
                    },
                    itemBuilder = { rating, selected ->
                        Row(
                            Modifier
                                .selectableRowColumnItemBackground(selected = selected)
                                .padding(vertical = 5.dp, horizontal = 10.dp)
                        ) {
                            Text(
                                rating.title,
                                style = MaterialTheme.typography.body1Light,
                                color = (if (selected) Color.White else Color.Black).toColor(),
                            )
                            if (rating != UsersFilter.Rating.All) {
                                Icon(
                                    Icons.Default.Star,
                                    contentDescription = null,
                                    tint = (if (selected) Color.White else Color.LightGray).toColor()
                                )
                            }
                        }
                    }
                )
            }
        }
        item {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceBetween,
                modifier = Modifier
                    .padding(horizontal = 20.dp, vertical = 10.dp)
                    .fillMaxWidth()
            ) {
                Text("With reviews")
                Switch(
                    checked = state.filter.withReviews,
                    onCheckedChange = {
                        listener(Msg.ToggleWithReviews)
                    }
                )
            }
        }
        item {
            ActionButton(
                style = ActionButtonStyle.White,
                onClick = {
                    listener(Msg.Clear)
                }) {
                Text("clear all")
            }
        }
        item {
            ActionButton(onClick = hide) {
                Text("Show")
            }
        }
    }
}

@Composable
fun <Item> SelectableRowColumnGrid(
    items: Collection<Item>,
    selectedIndex: Int,
    selectIndex: (Int) -> Unit,
    itemBuilder: @Composable (Item, Boolean) -> Unit,
    modifier: Modifier = Modifier,
    spacing: Dp = 9.dp,
) {
    FlowRow(
        modifier = modifier,
        mainAxisSpacing = spacing,
        crossAxisSpacing = spacing,
    ) {
        items.forEachIndexed { index, item ->
            Control(
                onClick = {
                    if (selectedIndex != index) {
                        selectIndex(index)
                    }
                }, content = {
                    itemBuilder(item, index == selectedIndex)
                }
            )
        }
    }
}

@SuppressLint("ComposableModifierFactory")
@Composable
private fun Modifier.selectableRowColumnItemBackground(selected: Boolean) =
    RoundedCornerShape(100).let { shape ->
        borderKMM(
            width = (if (selected) 0.0 else 1.5).dp,
            color = Color.LightGray,
            shape = shape
        )
            .backgroundKMM(
                if (selected) Color.Green else Color.Transparent,
                shape = shape
            )
    }